File: LeetCode/380. Insert Delete GetRandom O(1).js
// https://leetcode.com/problems/insert-delete-getrandom-o1/description/

class RandomizedSet {
  constructor() {
    this.map = new Map();
    this.list = [];
  }

  insert(val) {
    if (this.map.has(val)) return false;

    this.list.push(val);
    this.map.set(val, this.list.length - 1);

    return true;
  }

  remove(val) {
    if (!this.map.has(val)) return false;

    const index = this.map.get(val);

    this.list[index] = this.list[this.list.length - 1];
    this.map.set(this.list[index], index);

    this.map.delete(val);
    this.list.pop();
    return true;
  }

  getRandom() {
    const size = this.list.length;
    const randomIndex = Math.floor(Math.random() * size);

    return this.list[randomIndex];
  }
}

File: LeetCode/380. Insert Delete GetRandom O(1).py
# https://leetcode.com/problems/insert-delete-getrandom-o1/description/
from random import random


class RandomizedSet:

    def __init__(self) -> None:
        self.list: list[int] = []
        self.hash: dict[int, int] = {}

    def insert(self, val: int) -> bool:
        if val in self.hash:
            return False

        self.list.append(val)
        self.hash[val] = len(self.list) - 1

        return True

    def remove(self, val: int) -> bool:
        if val not in self.hash:
            return False

        index = self.hash[val]

        self.list[index] = self.list[(len(self.list)) - 1]
        self.hash[self.list[index]] = index

        del self.hash[val]
        self.list.pop()

        return True

    def getRandom(self) -> int:
        index = int(random() * len(self.list))

        return self.list[index]
